lib/BridgedRoom.js

```diff
@@ -188,81 +190,172 @@ BridgedRoom.prototype.toEntry = function() {
     return entry;
 };
 
-BridgedRoom.prototype.onMatrixMessage = function(message) {
+BridgedRoom.prototype.onMatrixRedaction = async function(message) {
+    if (!this._slack_bot_token) return Promise.resolve();
+
+    const eventStore = this._main.getEventStore();
+    const event = await eventStore.getEntryByMatrixId(message.room_id, message.redacts);
+
+    // If we don't get an event then exit
+    if (event === null) {
+        log.debug("Could not find event to delete.");
+        return Promise.resolve();
+    }
+
+    var body = {channel: this._slack_channel_id,
+                ts: event.remoteEventId,
+                as_user: false};
+
+    const sendMessageParams = {
+        method: "POST",
+        json: true,
+        uri: "https://slack.com/api/chat.delete",
+        body: body,
+    };
+
+    if (this._slack_bot_token) {
+        sendMessageParams.headers = {
+            Authorization: 'Bearer ' + this._slack_bot_token
+        };
+    }
+
+    return await rp(sendMessageParams);
+};
+
+BridgedRoom.prototype.onMatrixEdit = async function(message) {
     if (!this._slack_webhook_uri && !this._slack_bot_token) return Promise.resolve();
 
-    return this._main.getOrCreateMatrixUser(message.user_id).then((user) => {
-        substitutions.matrixToSlack(message, this._main).then(body => {
-            const uri = (this._slack_bot_token) ? "https://slack.com/api/chat.postMessage" : this._slack_webhook_uri;
-
-            const sendMessageParams = {
-                method: "POST",
-                json: true,
-                uri: uri,
-                body: body,
-            };
-            if (this._slack_bot_token) {
-                sendMessageParams.headers = {
-                    Authorization: 'Bearer ' + this._slack_bot_token
-                };
-                // See https://api.slack.com/methods/chat.postMessage#authorship
-                sendMessageParams.body.as_user = false;
-                sendMessageParams.body.channel = this._slack_channel_id;
-            }
+    const eventStore = this._main.getEventStore();
+    const event = await eventStore.getEntryByMatrixId(message.room_id, message.content['m.relates_to'].event_id);
 
-            sendMessageParams.body.username = user.getDisplaynameForRoom(message.room_id);
+    // re-write the message so the matrixToSlack converter works as expected.
+    const new_message = JSON.parse(JSON.stringify(message));
+    new_message.content = message.content['m.new_content'];
 
-            const avatar_url = user.getAvatarUrlForRoom(message.room_id);
+    const body = await substitutions.matrixToSlack(new_message, this._main);
 
-            if (avatar_url && avatar_url.indexOf("mxc://") === 0) {
-                sendMessageParams.body.icon_url = this._main.getUrlForMxc(avatar_url);
-            }
+    const sendMessageParams = {
+        method: "POST",
+        json: true,
+        uri: "https://slack.com/api/chat.update",
```

Comment:
Reviewer note: We really need a slack client rather than these awful url-param object things.

Reply:
I mean sure, suggestions?

Reply:
Let's just fix it in a new PR